### PR TITLE
grub2: 2.06 -> unstable-2023-07-03

### PR DESCRIPTION
**Pull request by K900**
- build from latest git
- remove all the backports
- manually forward-port + vendor the hiddenentry patch (original link was dead btw)
- vendor the exact gnulib version grub wants

###### Description of changes

Our GRUB is very very broken. One could argue that that's just GRUB in general, but at least the version from master seems _less_ broken, and Arch is shipping it, and the giant pile of backports has become unmaintainable a long time ago.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by K900** [comment]
Result of `nixpkgs-review pr 240887` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>11 packages built:</summary>
  <ul>
    <li>grub2 (grub2_light)</li>
    <li>grub2.debug (grub2_light.debug)</li>
    <li>grub2_efi</li>
    <li>grub2_efi.debug</li>
    <li>grub2_pvgrub_image</li>
    <li>grub2_xen</li>
    <li>os-prober</li>
    <li>timeshift</li>
    <li>woeusb</li>
    <li>woeusb-ng</li>
    <li>woeusb-ng.dist</li>
  </ul>
</details>

**Comment by Et7f3** [comment]
fuse3 could then be used ?

**Comment by K900** [comment]
Possibly, I wanted to keep this change as small as possible to ensure we don't break things even more.

File: pkgs/tools/misc/grub/default.nix [hunk 1]
```diff
@@ -42,8 +40,12 @@ let
   canEfi = lib.any (system: stdenv.hostPlatform.system == system) (lib.mapAttrsToList (name: _: name) efiSystemsBuild);
   inPCSystems = lib.any (system: stdenv.hostPlatform.system == system) (lib.mapAttrsToList (name: _: name) pcSystems);
 
-  version = "2.06";
-
+  gnulib = fetchFromSavannah {
+    repo = "gnulib";
+    # NOTE: keep in sync with bootstrap.conf!
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/grub/default.nix:45]: Can we check this with a grep or so?

File: pkgs/tools/misc/grub/default.nix [hunk 2]
```diff
@@ -52,301 +54,17 @@ assert !(efiSupport && xenSupport);
 
 stdenv.mkDerivation rec {
   pname = "grub";
-  inherit version;
+  version = "unstable-2023-06-23";
 
-  src = fetchurl {
-    url = "mirror://gnu/grub/grub-${version}.tar.xz";
-    sha256 = "sha256-t56kSvkbk9F80/6Ava5u1DdwZ4qaWuGSzOqAPrtlfuE=";
+  src = fetchFromSavannah {
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/grub/default.nix:59]:
```suggestion
  # https://git.savannah.gnu.org/cgit/grub.git/
  src = fetchFromSavannah {
```

File: pkgs/tools/misc/grub/default.nix [hunk 3]
```diff
@@ -52,301 +54,17 @@ assert !(efiSupport && xenSupport);
...
-      url = "https://git.savannah.gnu.org/cgit/grub.git/patch/?id=1514678888595ef41a968a0c69b7ff769edd1e9c";
-      sha256 = "sha256-tgAEoAtaNKJjscjMFkXXiVn59Pa4c+NiQ3iVW6CMrpo=";
-    })
-
-    # fix incompatibility with e2fsprogs 1.47+
-    (fetchpatch {
-      url = "https://git.savannah.gnu.org/cgit/grub.git/patch/?id=7fd5feff97c4b1f446f8fcf6d37aca0c64e7c763";
-      sha256 = "sha256-pejn1bJkC7XnT2ODaxeERHUrMOONoBV6w0wF2Z2ZKWI=";
-    })
+    ./add-hidden-menu-entries.patch
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/grub/default.nix:67]: I am curious about this patch

File: pkgs/tools/misc/grub/default.nix (hunk 1 above)
Comment (K900) [reply pkgs/tools/misc/grub/default.nix:45]: We probably should, yeah. I'll see if there's an easy way to do it.

File: pkgs/tools/misc/grub/default.nix (hunk 3 above)
Comment (K900) [reply pkgs/tools/misc/grub/default.nix:67]: It's used by the live ISO to hotkey switching between text and graphical modes.

File: pkgs/tools/misc/grub/default.nix (hunk 2 above)
Comment (Et7f3) [reply pkgs/tools/misc/grub/default.nix:59]: We don't put similar comment on other fetchFromSavannah or fetchFromGithub + it should be easier queryable with .src.url ?

**Comment by adamcstephens** [comment]
When switching I received an error about locale (not present on previous config)

```
updating GRUB 2 menu...
installing the GRUB 2 boot loader into /boot...
Installing for x86_64-efi platform.
/nix/store/kgyhiwrnv6vf17gv5rdyav4apgi6bayh-grub-unstable-2023-06-23/sbin/grub-install: warning: cannot open directory `/nix/store/kgyhiwrnv6vf17gv5rdyav4apgi6bayh-grub-unstable-2023-06-23/share/locale': No such file or directory.
Installation finished. No error reported.
```

File: pkgs/tools/misc/grub/default.nix (hunk 1 above)
Comment (K900) [reply pkgs/tools/misc/grub/default.nix:45]: Done.

File: pkgs/tools/misc/grub/default.nix (hunk 2 above)
Comment (K900) [reply pkgs/tools/misc/grub/default.nix]: Don't really see the point tbh.

**Review by adamcstephens — APPROVED** [review]
Tested installation using a pure EFI (nodev) host and a legacy MBR/BIOS/CSM host.

**Comment by samueldr** [comment]
> Our GRUB is very very broken.

Do you have a list of things that updating addresses? I definitely don't doubt that fact. I'm curious what the direct intention is here.

I'll test on some systems (importantly AArch64) with my setup and report back if for my use-case there are issues.

File: pkgs/tools/misc/grub/default.nix [hunk 4]
```diff
@@ -42,8 +40,35 @@ let
...
+  src = fetchFromSavannah {
+    repo = "grub";
+    rev = "6425c12cd77ad51ad24be84c092aefacf0875089";
+    hash = "sha256-PSCa993Reph6w9+leE4a/9E6vIALdOqU3FZEPwasFyk=";
+  };
 
+  # HACK: the translations are stored on a different server,
+  # not versioned and not included in the git repo, so fetch them
+  # and hope they don't change often
+  locales = runCommand "grub-locales" {
```
Comment (2xsaiko) [inline pkgs/tools/misc/grub/default.nix:59]: would this work with runCommandNoCC?

Comment (K900) [reply pkgs/tools/misc/grub/default.nix:59]: There is no `runCommandNoCC` anymore, `runCommand` is always `NoCC`.

Comment (2xsaiko) [reply pkgs/tools/misc/grub/default.nix:59]: Oh, that's good to know!

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by K900** [comment]
Let's see what happens.

**Comment by ius** [comment]
Allegedly this happened: https://github.com/NixOS/nixpkgs/issues/245101

**Pull request merged**